internal/util/test/init.go

```diff
@@ -0,0 +1,39 @@
+package test
+
+import (
+	"fmt"
+	"path"
+	"path/filepath"
+	"runtime"
+)
+
+var (
+	kongRBACsKustomize        = "config/rbac/"
+	kongGatewayRBACsKustomize = "config/rbac/gateway"
+	kongCRDsRBACsKustomize    = "config/rbac/crds"
+
+	kongCRDsKustomize = "config/crd/"
+)
+
+// init initializes kustomize paths relative to the repo root directory.
+func init() {
+	root, err := getRepoRoot()
+	if err != nil {
+		panic(err)
+	}
+
+	kongCRDsKustomize = filepath.Join(root, kongCRDsKustomize)
+
+	kongRBACsKustomize = filepath.Join(root, kongRBACsKustomize)
+	kongGatewayRBACsKustomize = filepath.Join(root, kongGatewayRBACsKustomize)
+	kongCRDsRBACsKustomize = filepath.Join(root, kongCRDsRBACsKustomize)
+}
```

Comment:
I agree that it would be better to not use their initial values to get the final value but using functions like suggested above will make the tests not fail immediately (they will only fail when called first time) if someone was to rename/remove those directories, thus extending the time in which the developer gets feedback. Let me suggest something slightly different based on your suggestion.

Reply:
9087a161b

Reply:
>using functions like suggested above will make the tests not fail immediately (they will only fail when called first time) if someone was to rename/remove those directories, thus extending the time in which the developer gets feedback

I'm not sure if this is true. 🤔 As I understand, these are called as soon as this package gets imported, the same as `init()` because they're used to assign values to the global variables. I agree that ensuring the directories exist can improve DevX and it makes sense. How about meeting in-between (use `ensureDirExists` and do not use intermediary consts 😅 )?

```go
var (
	kongRBACsKustomize        = initKongRBACsKustomizePath()
	kongGatewayRBACsKustomize = initKongGatewayRBACsKustomizePath()
	kongCRDsRBACsKustomize    = initKongCRDsRBACsKustomizePath()
	kongCRDsKustomize         = initCRDsKustomizePath()
)

func initKongRBACsKustomizePath() string {
	dir := filepath.Join(lo.Must(getRepoRoot()), "config/rbac/")
	ensureDirExists(dir)
	return dir
}
func initKongGatewayRBACsKustomizePath() string {
	dir := filepath.Join(lo.Must(getRepoRoot()), "config/rbac/gateway")
	ensureDirExists(dir)
	return dir
}
func initKongCRDsRBACsKustomizePath() string {
	dir := filepath.Join(lo.Must(getRepoRoot()), "config/rbac/crds")
	ensureDirExists(dir)
	return dir
}
func initCRDsKustomizePath() string {
	dir := filepath.Join(lo.Must(getRepoRoot()), "config/crd/")
	ensureDirExists(dir)
	return dir
}
```

I verified it fails the same way as the `init()` version just on startup:
```shell
make test.integration.isolated.dbless
INFO: container environment ready DOCKER=(Docker version 24.0.6, build ed223bc) KIND=(kind v0.19.0 go1.21.0 darwin/arm64)
(cd third_party && go mod tidy && \
                GOBIN=/Users/grzegorz.burzynski@konghq.com/Repos/kubernetes-ingress-controller/bin go generate -tags=third_party ./go-junit-report.go )
KONG_CLUSTER_VERSION="v1.28.0" \
                TEST_KONG_HELM_CHART_VERSION="2.31.0" \
                TEST_DATABASE_MODE="off" \
                GOFLAGS="-tags=integration_tests" \
                KONG_CONTROLLER_FEATURE_GATES="GatewayAlpha=true" \
                go test  \
                -v \
                -timeout "45m" \
                -parallel 10 \
                -race \
                -covermode=atomic \
                -coverpkg=./pkg/...,./internal/... \
                -coverprofile=coverage.dbless.out \
                ./test/integration/isolated -args --parallel  | \
        /Users/grzegorz.burzynski@konghq.com/Repos/kubernetes-ingress-controller/bin/go-junit-report -iocopy -out /dev/null -parser gotest
panic: stat /Users/grzegorz.burzynski@konghq.com/Repos/kubernetes-ingress-controller/config/rb: no such file or directory

goroutine 1 [running]:
github.com/kong/kubernetes-ingress-controller/v3/internal/util/test.ensureDirExists({0xc000199980, 0x52})
        /Users/grzegorz.burzynski@konghq.com/Repos/kubernetes-ingress-controller/internal/util/test/init.go:44 +0x158
github.com/kong/kubernetes-ingress-controller/v3/internal/util/test.initKongRBACsKustomizePath()
        /Users/grzegorz.burzynski@konghq.com/Repos/kubernetes-ingress-controller/internal/util/test/init.go:22 +0xf4
github.com/kong/kubernetes-ingress-controller/v3/internal/util/test.init()
        /Users/grzegorz.burzynski@konghq.com/Repos/kubernetes-ingress-controller/internal/util/test/init.go:14 +0xd4
FAIL    github.com/kong/kubernetes-ingress-controller/v3/test/integration/isolated      1.535s
FAIL

```

Reply:
Alright, I was focused on the `func init()` part and I didn't realise that initialising `var`s with return values of functions would yield a similar result 🤦 

ccb19dfb9